ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,241 @@
+    this.time = time;
+    this.diskIOScheduler = diskIOScheduler;
+  }
+
+  // TODO remove this once #541 is merged
+  public static String getServiceId(StoreKey key) {
+    return key.getID().substring(0, 1);
+  }
+
+  // TODO remove this once #541 is merged
+  public static String getContainerId(StoreKey key) {
+    return key.getID().substring(key.getID().length() - 1);
+  }
+
+  @Override
+  public Pair<Long, Long> getValidDataSize(TimeRange timeRange) throws StoreException {
+    Pair<Long, Map<String, Long>> logSegmentValidSizeResult = getValidDataSizeByLogSegment(timeRange);
+    Long allLogSegmentValidDataSize = 0L;
+    for (Long value : logSegmentValidSizeResult.getSecond().values()) {
+      allLogSegmentValidDataSize += value;
+    }
+    return new Pair<>(logSegmentValidSizeResult.getFirst(), allLogSegmentValidDataSize);
+  }
+
+  @Override
+  public Pair<Long, Map<String, Long>> getValidDataSizeByLogSegment(TimeRange timeRange) throws StoreException {
+    long referenceTimeInSecs = timeRange.getEndTimeInSecs();
+    Map<String, Long> logSegmentValidSizeMap = collectValidDataSizeByLogSegment(referenceTimeInSecs * Time.MsPerSec);
+    return new Pair<>(referenceTimeInSecs, logSegmentValidSizeMap);
+  }
+
+  @Override
+  public Map<String, Map<String, Long>> getValidDataSize() throws StoreException {
+    long referenceTimeInMs = time.milliseconds();
+    return collectValidDataSizeByContainer(referenceTimeInMs);
+  }
+
+  /**
+   * Walk through the entire index and collect valid data size information per container.
+   * @param referenceTimeInMs the reference time to be used to decide whether or not a blob is valid
+   * @return a nested {@link Map} of serviceId to containerId to valid data size
+   */
+  private Map<String, Map<String, Long>> collectValidDataSizeByContainer(long referenceTimeInMs)
+      throws StoreException {
+    Offset previousSegmentEndOffset = null;
+    Offset indexStartOffset = index.getStartOffset();
+    Map<String, Map<String, Long>> containerValidSizeMap = new HashMap<>();
+    for (IndexSegment indexSegment : index.getIndexSegments().values()) {
+      diskIOScheduler.getSlice(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, BlobStoreStats.IO_SCHEDULER_JOB_ID, 1);
+      List<MessageInfo> messageInfos = new ArrayList<>();
+      try {
+        indexSegment.getEntriesSince(null, new FindEntriesCondition(Integer.MAX_VALUE), messageInfos, new AtomicLong(0));
+      } catch (IOException e) {
+        throw new StoreException("I/O error while collecting valid data size information", e, StoreErrorCodes.IOError);
+      }
+      Offset backwardSearchEndOffset =
+          previousSegmentEndOffset == null ? indexSegment.getStartOffset() : previousSegmentEndOffset;
+      FileSpan backwardSearchSpan = new FileSpan(indexStartOffset, backwardSearchEndOffset);
+      for (MessageInfo messageInfo : messageInfos) {
+        populateContainerMap(containerValidSizeMap, referenceTimeInMs, messageInfo, backwardSearchSpan);
```

Comment:
why can't you pass the entire list of MessageInfos to populateContainerMap and populate for every MessageInfo